### PR TITLE
Validators

### PR DESCRIPTION
**Pull request by xRyuki-i**
I made the changes that I was asked to do:

1) Improvising the error messages.
2)Refactoring the min and max validators to check the value itself.
3)Refactoring the date validator to validate a date object instead of a string
4)Fixing the lint errors

Also, I have added the minDate and maxDate validator. I've tested them successfully.

### CODE REVIEW

File: src/validators.test.ts [hunk 1]
```diff
@@ -66,3 +68,131 @@ describe('length', () => {
...
+    expect(maxlength(3, errorMsg)("suku")).toEqual(errorMsg);
+  });
+})
+
+ // min validator
+describe('min', () => {
+  it('returns error', () => {
+    const val = 12
+    const minVal = 15;
+    const expectedError = `The entered value: ${val} should be at greater than ${minVal}`;
```
Comment (ludbek) [inline src/validators.test.ts]: `Value ${val} should be greater than ${minVal}` would look better. What is your thought on this?

File: src/validators.test.ts [hunk 2]
```diff
@@ -66,3 +68,131 @@ describe('length', () => {
...
+    expect(min(6, errorMsg)(5)).toEqual(errorMsg);
+  });
+})
+
+ // max validator
+describe('max', () => {
+  it('returns error', () => {
+    const val = 15;
+    const maxVal = 12;
+    const expectedError = `The entered value: ${val} should be smaller than ${maxVal}`;
```
Comment (ludbek) [inline src/validators.test.ts]: Similar to above.

File: src/validators.test.ts [hunk 3]
```diff
@@ -66,3 +68,131 @@ describe('length', () => {
...
+  });
+  it('considers errorMsg', () => {
+    const maxVal = 3;
+    const errorMsg = `Number smaller than ${maxVal} is expected`;
+    expect(max(maxVal, errorMsg)(5)).toEqual(errorMsg);
+  });
+})
+
+ // date validator
+describe('validateDate', () => {
```
Comment (ludbek) [inline src/validators.test.ts]: We might not need this. The date `decoder` will make this validation. I just you to move this logic to the date validator. At the moment, its implementation is wrong. It should take an ISO string and
convert it to date and validate at the same time.

File: src/validators.ts [hunk 4]
```diff
@@ -96,11 +99,171 @@ export function length(
...
+  const day = parseInt(parts[2], 10);
+  const month = months.indexOf(parts[1]) + 1;
+  const year = parseInt(parts[3], 10);
+  
+  const dateValue:number[] = [day, month, year];
+  return dateValue;
+}
+
+ // dateFormat
+export function validateDate(
```
Comment (ludbek) [inline src/validators.ts]: We need some love to the date decoder. Its located here https://github.com/ludbek/validatex/blob/master/src/decoders.ts#L170

Its implementation is wrong at the moment. It should take an unknown value and convert it to a date. The expected unknown values will be either a string or a date itself. Try moving these logics to there.

Also we might not need to parse the date string manually. Javascript is already capable of parsing an ISO format dates. Checkout https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/parse

File: src/validators.ts [hunk 5]
```diff
@@ -96,11 +99,171 @@ export function length(
...
+  }
+}
+
+ // minDate
+export function minDate(
+  option : Date,
+  errorMsg?: CustomErrMsg<Date, undefined>,
+) {
+  return (val: Date, context?: Context): string | undefined => {
+    const minimum = parseDate(option);
```
Comment (ludbek) [inline src/validators.ts]: Use `Date.parse` instead.

File: src/validators.ts [hunk 6]
```diff
@@ -96,11 +99,171 @@ export function length(
...
+  }
+}
+
+ // maxDate
+export function maxDate(
+  option : Date,
+  errorMsg?: CustomErrMsg<Date, undefined>,
+) {
+  return (val: Date, context?: Context): string | undefined => {
+    const maximum = parseDate(option);
```
Comment (ludbek) [inline src/validators.ts]: Same as above

File: src/validators.ts [hunk 7]
```diff
@@ -96,11 +99,171 @@ export function length(
...
+    if (validateDate()(val) !== undefined ){
+      return getError({ errorMsg, defaultErrorMsg, val, context, option:undefined })
+    }
+
+    const date = parseDate(val);
+    const day = date[0];
+    const month = date[1];
+    const year = date[2];
+
+    if (year < maxYear){
```
Comment (ludbek) [inline src/validators.ts]: We won't need this. Read about the date type in Javascript and see how we can compare 2 instances of dates.

File: src/decoders.ts [hunk 8]
```diff
@@ -167,12 +167,15 @@ const dBoolean = decoder<boolean, unknown>({
   defaultParser: identity,
 });
 
+ // moving the dateValidator logic to here
 const date = decoder<Date, unknown>({
-  typeGuard: (val: unknown): val is Date => val instanceof Date,
-  getDefaultErrorMsg: (val) => `Expected Date but got ${serializeType(val)}`,
+  typeGuard: (val: unknown): val is Date => val instanceof Date && val.toDateString() != "Invalid Date",
```
Comment (xRyuki-i) [inline src/decoders.ts]: The lint is asking to change double equals to triple but the validation works with double equals only. What should be done?

Comment (ludbek) [reply src/decoders.ts]: Checkout the difference between === and == . You will get your answer

File: src/validators.test.ts [hunk 9]
```diff
@@ -66,3 +68,131 @@ describe('length', () => {
...
+  it('considers errorMsg', () => {
+    const maxVal = 3;
+    const errorMsg = `Number smaller than ${maxVal} is expected`;
+    expect(max(maxVal, errorMsg)(5)).toEqual(errorMsg);
+  });
+})
+
+ // date validator
+// describe('validateDate', () => {
+//   it('returns error', () => {
```
Comment (ludbek) [inline src/validators.test.ts:157]: If don't need this code, it's better to remove it all together.

File: src/decoders.ts (hunk 8 above)
Comment (xRyuki-i) [reply src/decoders.ts]: I know that === checks the type and the value whereas == only checks the value.

**Pull request closed without merging**